Wrap non-latin characters in a span specifying working font

adam-james-v opened the pull request:
Fixes: #38753

Please keep in mind that this bug will only show up in rendered table cards, on Slack. All other static-viz chart types render as SVGs via Batik.

CSSBox seems to have a bug where font fallback doesn't work properly. This is noticeable when a font does not contain glyphs that are present in the string being rendered. For example, Lato does not have international characters (Cyrillic, Hangul, etc.), so the rendered version of tables (that show up in Slack messages) will not render properly, making the card unreadable.

Since this appears to be a downstream bug, I've opted to work around this limitation by wrapping strings containing non-lato characters in a `<span>` that specifies the font family to be sans-serif, which should properly fall back to a font that contains the glyphs and thus render legibly. 

This is done with a postwalk over the Hiccup data structure, and only attempts to wrap strings inside `#{:div :span :td :th :tr :table :p :tbody :thead}`. If a string contains any glyphs not present in the Lato font, the entire string is wrapped. This is to keep mixed strings (Latin + non-latin chars/glyphs) visibly cohesive.

### Why doesn't the fallback work as we would expect, can't we configure things to work properly?

We load the Lato font directly into our GraphicsEnvironment using `java.awt.Font/createFont`
https://github.com/metabase/metabase/blob/88a31b41217f14b6ead98829c53cd5068b5a4ba8/src/metabase/pulse/render/style.clj#L102

We do that because it's used in the SVG rendering to give charts the proper font+style. If we do not load the font in, our SVG charts render slightly incorrectly (font weight is not correct, for example).

When you create a font in this way, it is a 'physical font', and will not fallback to another font when glyphs don't exist. Logical fonts DO, but we cannot create logical fonts, they are defined by the JRE. We can derive new fonts from logical fonts, but that copies *attributes* such as size and weight but not glyphs.

It *is* possible to have fonts fall back appropriately if you rely on the system's font setup. That is, if you do NOT register Lato, and you DO have lato installed on your machine (at least for me it worked on MacOS), then fallbacks seem to work. I did not go with this approach because I want static-viz renders to still work 'out of the box', so a user shouldn't have to install Lato to get proper renders.

For now, I figured it's worth trying this solution over using Noto for 2 reasons:
- we can keep Lato, which has been the decided font style for the app for some time (this keeps things consistent where possible)
- the Noto font containing all glyphs is quite a large font (>20mb, I think) and it would be nice to avoid bundling that if we can.

The fallback clearly isn't working in the first render, which is what the render will look like without this change.
And The bottom render is after this wrapping was added:

<img width="333" alt="image" src="https://github.com/metabase/metabase/assets/21064735/3b875592-7721-40a0-9cb8-5f24bc6beff3">

adam-james-v committed:
Wrap non-latin characters in a span specifying working font

Fixes: #38753

CSSBox seems to have a bug where font fallback doesn't work properly. This is noticeable when a font does not contain
glyphs that are present in the string being rendered. For example, Lato does not have many international characters,
so the rendered version of tables (that show up in Slack messages) will not render properly, making the card
unreadable.

Since this appears to be a downstream bug, I've opted to work around this limitation by wrapping any non-latin
characters in a <span> that specifies the font family to be sans-serif, which should contain the glyphs to properly
render.

This leaves Lato in place for other characters.

For now, I figured it's worth trying this solution over using Noto for 2 reasons:
- we can keep Lato, which has been the decided font style for the app for some time (this keeps things consistent
where possible)
- the Noto font containing all glyphs is quite a large font (>20mb, I think) and it would be nice to avoid bundling
that if we can.

piranha commented:
It is kind of weird, I've checked [Lato's wikipedia page](https://en.wikipedia.org/wiki/Lato_(typeface)) and it says it supports Cyrillic alphabet, but on Google Fonts it does not. I wonder if it makes sense downloading [from their site](https://www.latofonts.com/lato-free-fonts/) and trying that out?

adam-james-v commented:
> It is kind of weird, I've checked [Lato's wikipedia page](https://en.wikipedia.org/wiki/Lato_(typeface)) and it says it supports Cyrillic alphabet, but on Google Fonts it does not. I wonder if it makes sense downloading [from their site](https://www.latofonts.com/lato-free-fonts/) and trying that out?

Hmm... even if it does support Cyrillic alphabet, that's not sufficient coverage, I don't think. For instance, Hangul (Korean characters, pardon my ignorance if this isn't the right way to refer to them), I don't think is covered anyway.

camsaul reviewed (dismissed):
1. Could we just add the `font-family` style to the top-level `<body>` tag instead of trying to find each non-ASCII character and wrapping each one in its own `<span>`? That way the entire document defaults to it and you don't need to do the risky regex search-and-replace? You could actually do that in the `render-html-to-png` function below this one, we're already setting a few other things like `margin: 0` there so you can just add it to that. Don't forget `font-family` can have multiple fonts as fallbacks so you can specify whatever fancy font first and then `sans-serif` as the fallback

2. Splicing in raw HTML like this has me extremely itchy, if you actually do need to do something like this I think it would have been better to walk the Hiccup form in the `render-html-to-png` function below and wrap element bodies as needed band output modified Hiccup rather than try to search and replace in the compiled HTML. That way we don't have to worry about something like `<a href="example.com/jalapeño">` getting replaced with `<a href="example.com/jalape<span style=\"font-family: sans-serif;\">ñ</span>o">`. Then you can also test this stuff much more easily

   But of course it's even better if you can just specify the font family in the `<body>` and don't have to worry about walking anything in the first place.
   
3. This really needs some tests

4. Something like this really needs some comments explaining why you're doing what you're doing here, in a year someone is going to come in here and have to spend 10 minutes trying to figure out why in the world we're doing this

camsaul commented:
Ok actually I just saw your PR description where you mentioned that CSS box font fallback doesn't work properly and Ngoc's upstream issue https://github.com/radkovo/CSSBox/issues/85 ...

Have you looked at submitting a patch to CSSBox to fix this instead of the risky hack?

I've looked at the code for a few minutes and I found the string drawing code

https://github.com/radkovo/CSSBox/blob/c3e3e0832867da4d3a8d36df13b3a203ee2c5671/src/main/java/org/fit/cssbox/awt/GraphicsRenderer.java#L423-L438

It's using AWT which I'm pretty sure has a built-in notion of fallback fonts, maybe there's a way we can configure things to fall back to Noto or whatever else when it's rendering stuff. Maybe https://docs.oracle.com/javase/8/docs/technotes/guides/intl/fontconfig.html is worth reading 

Also we are already bundling Noto in the Docker image 

https://github.com/metabase/metabase/blob/8cda1be2676ed314edb8464db86a8f9ab43574d8/Dockerfile#L37

so it's already there and the JRE should have access to it

camsaul commented:
Here [this seems relevant](https://docs.oracle.com/javase/8/docs/technotes/guides/intl/font.html)

> Users can add a limited number of supported TrueType and Type 1 physical fonts as fallback fonts for logical fonts used in Java 2D rendering by installing them in the lib/fonts/fallback directory within the JRE.

That's worth trying. If it fixes things we could easily symlink Noto to that directory when we build our Docker image so the JRE will be able to use it as a fallback font when it renders AWT stuff

adam-james-v commented:
Thanks for the comments, Cam. You're right about the drawbacks here.
I was also experimenting locally. I don't know exactly the reason yet, but it turns out if I specify "Lato Regular", the rendering issue goes away. So, I'll be undoing this approach anyway, and trying something along the lines of what you say, related to making sure our java.awt font stuff is properly configured.

piranha commented:
> Hmm... even if it does support Cyrillic alphabet, that's not sufficient coverage, I don't think.

Yeah, but then it'll cover much more, especially for alphabets where it's very easy to notice that the font is different (Cyrillic and Greek mainly). You have no idea how _bad_ it looks when part of your string is one font and part is another.

camsaul commented:
Honestly, is that a real problem we expect people to see a lot of -- mixed fonts in the same sentence because the sentence is mixing different languages with different writing systems? I would expect most stuff to be in just one language.

adam-james-v committed:
stub installed fonts test

adam-james-v committed:
typo

piranha commented:
> sentence is mixing different languages with different writing systems

This is extremely common outside of English-speaking world.

adam-james-v committed:
Do wrapping, but now per-string, and in Clojure data not html string

I've decided that a reasonable solution is to still wrap strings containing non-lato characters. But it's not done
with str/replace to the html string but rather in a postwalk over the Hiccup data prior to rendering.

Seems like a decent compromise of issues without patching CSSBox or fixing upstream (may be good to do, but will take
longer to get a fix in).

adam-james-v committed:
add test checking that glyphs render correctly

adam-james-v committed:
Add a test that directly checks the wrapping fn

adam-james-v committed:
Change the string to keep the linter quiet

adam-james-v commented on test/metabase/pulse/render/png_test.clj:
```diff
@@ -13,9 +20,64 @@
+      ;; The broken-render's width is around 17px. It is the width of 2 `[?]` charaters
+      ;; We assert that the working render is wider based on the assumption (verified manually by
+      ;; actually looking at the rendered images) that the correctly rendered glyphs are wider.
+      (is (< (.getWidth broken-render) (.getWidth working-render))))))
```
For a visual ref, here's the broken/working render:

<img width="835" alt="image" src="https://github.com/metabase/metabase/assets/21064735/91690d3b-6b5e-4ca8-a778-3ec92def2160">

piranha commented on src/metabase/pulse/render/png.clj:
```diff
@@ -73,16 +74,66 @@
+
+#_
+(def ^:private non-latin-regex
+  (re-pattern "[^A-Za-z0-9!@#$%^&*()_+\\-=\\[\\]{}|;:'\",.<>?/`~\\s\\t\\n\\r]"))
```
I still think this is too narrow; Lato supports cyrillic and greek chars.

piranha replied:
disregard that comment, I did not notice `#_` :)

adam-james-v replied:
It's left there as a commented out form. Perhaps I should just remove it.

adam-james-v committed:
Change how we check if string can be rendered to faster method, new Lato Font

With Sanya's help, the way I check if a given string is renderable with Lato is now faster.

Also use the full Lato font, not just the 'latin' lato so we can cover more chars

sansolovyov reviewed (approved):
Aside from tests failing if you run only them because no fonts registered - works nicely!

adam-james-v committed:
change lato so that it loads the fn even in a fresh test run

camsaul commented on src/metabase/pulse/render/png.clj:
```diff
@@ -73,16 +74,65 @@
+  [^Font font s]
+  (neg? (.canDisplayUpTo font s)))
+
+(def ^:private get-lato
```
If you add `^{:arglists '(^java.awt.Font [])}` metadata on this then you can use `(get-lato)` directly in Java interop without having to worry about type hinting it

piranha replied:
Isn't it ok to put `def ^:private get-lato ^Font [...]` directly? `:arglists` are more for varargs thingie I guess and it could be simpler here?

camsaul commented on src/metabase/pulse/render/png.clj:
```diff
@@ -73,16 +74,65 @@
+
+(defn- lato-can-render?
+  [s]
+  (let [lato (get-lato)]
+    (when lato
```
You can use `when-let` here

camsaul commented on test/metabase/pulse/render/png_test.clj:
```diff
@@ -13,9 +20,66 @@
+      (png/render-html-to-png width)
+      bytes->image))
+
+(deftest wrap-non-lato-characters-test
```
Can you make this test `^:parallel`?
